### PR TITLE
Run UI E2E Tests on Saucelabs using Travis

### PR DESCRIPTION
**Pull request by tombee**
This adds the initial rigging to run E2E UI tests on Travis using Saucelabs.

I still need to figure out a nice workflow for making iterating on tests easier than building the entire application and then running the `hack/ui-test.sh` script.  I think improvements to the development workflow will become clearer as we start developing tests during UI development.

### CODE REVIEW

**Comment by ehazlett** [comment]
How long does it take to run the UI tests on Sauce?

File: ui/e2e/Dockerfile.sauceconnect [hunk 1]
```diff
@@ -0,0 +1,14 @@
+FROM ubuntu
```
Comment (tombee) [inline ui/e2e/Dockerfile.sauceconnect]: Wanted to make this `alpine` based, but I couldn't get it to work as it's seemingly Oracle Java 8 based and suffers glibc linking issues documented in: https://github.com/gliderlabs/docker-alpine/blob/master/docs/caveats.md#incompatible-binaries 😢

**Comment by tombee** [comment]
@ehazlett 

> How long does it take to run the UI tests on Sauce?

I've just run it with `time` and the results are:

```
time make check build test image
...
real	2m36.322s
```

```
time ./hack/ui-test.sh
...
real	1m24.707s
```

**Comment by tombee** [comment]
Rebased

File: ui/e2e/Dockerfile.sauceconnect (hunk 1 above)
Comment (ehazlett) [inline ui/e2e/Dockerfile.sauceconnect]: :(

Comment (ehazlett) [inline ui/e2e/Dockerfile.sauceconnect]: What about `java:openjdk-8-jdk-alpine` or one of the other alpine tags for the java image?

File: ui/e2e/nightwatch.conf.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
+require("babel-core/register");
```
Comment (ehazlett) [inline ui/e2e/nightwatch.conf.js]: We're getting too much yoloscript in this repo ;) jk jk

File: ui/e2e/nightwatch.json [hunk 3]
```diff
@@ -0,0 +1,67 @@
+{
```
Comment (ehazlett) [inline ui/e2e/nightwatch.json]: Could we add some quick docs for us non-UI folks to understand what this does?  i.e. how nightwatch, etc all work together.  You don't have to go too deep we can RTFM just a high level.

File: ui/e2e/Dockerfile.sauceconnect (hunk 1 above)
Comment (tombee) [inline ui/e2e/Dockerfile.sauceconnect]: Yeah, I'd tried `java:openjdk-8-jdk-alpine`, but sadly that doesn't work either :cry:

File: ui/e2e/nightwatch.json (hunk 3 above)
Comment (tombee) [inline ui/e2e/nightwatch.json]: Yeah, will do.. I was going to start adding as I go, I haven't really figured out the ideal "development workflow" for developing and iterating on tests with saucelabs yet.

**Comment by tombee** [comment]
> this looks fine to me. we can do this in a follow up but it would probably be nice to wrap all of these ./hack/... scripts to be make targets. So we can just make test or make test-ui etc.

Cool okay!  I'd started off by having it in the Makefile, but with all the setup/teardown it was starting to get quite bloated, so I'd kept the `hack/ui-test.sh` script separate.

File: project/testing.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+describe("App", () => {
+  it("matches snapshot", () => {
+    const wrapper = shallow(<App />);
+    expect(wrapper).toMatchSnapshot();
+  });
+});
+```
+
+
+### End-to-End Tests
```
Comment (tombee) [inline project/testing.md]: This process for running end-to-end tests isn't going to be ideal for running locally, or when developing/iterating/troubleshooting tests.  I've just documented how it's working at the moment, just to get travis-ci running.

I'll be refining this process as we go, if that's alright?

**Comment by tombee** [comment]
@ehazlett I've updated this with some docs on the current test strategy.  There's still more refinement to be done, but I think that it makes sense to do that as we start building out our test suite and figure out what works for us.

**Comment by ehazlett** [comment]
LGTM

**Pull request merged**